File: crates/medousa-sdk/src/transport.rs
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use serde::de::DeserializeOwned;

use crate::error::SdkError;

/// Object-safe HTTP transport for the Medousa daemon API.
pub trait Transport: Send + Sync {
    fn get_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>>;

    fn post_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
        body: serde_json::Value,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>>;

    fn delete_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>>;

    fn put_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
        body: serde_json::Value,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>>;
}

pub async fn decode<T: DeserializeOwned>(value: serde_json::Value) -> Result<T, SdkError> {
    serde_json::from_value(value).map_err(Into::into)
}

#[derive(Clone, Default)]
pub struct HttpTransport {
    client: reqwest::Client,
}

impl HttpTransport {
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::new(),
        }
    }

    pub fn with_client(client: reqwest::Client) -> Self {
        Self { client }
    }

    fn url(base_url: &str, path: &str) -> String {
        let base = base_url.trim_end_matches('/');
        let path = if path.starts_with('/') {
            path.to_string()
        } else {
            format!("/{path}")
        };
        format!("{base}{path}")
    }

    async fn request(
        client: reqwest::Client,
        method: reqwest::Method,
        url: String,
        body: Option<serde_json::Value>,
    ) -> Result<serde_json::Value, SdkError> {
        let mut builder = client.request(method, url);
        if let Some(body) = body {
            builder = builder.json(&body);
        }
        let response = builder.send().await?;
        let status = response.status();
        let text = response.text().await?;
        if !status.is_success() {
            return Err(SdkError::Http(format!("{status}: {text}")));
        }
        if text.trim().is_empty() {
            return Ok(serde_json::Value::Null);
        }
        Ok(serde_json::from_str(&text)?)
    }
}

impl Transport for HttpTransport {
    fn get_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>> {
        let url = Self::url(base_url, path);
        let client = self.client.clone();
        Box::pin(async move {
            Self::request(client, reqwest::Method::GET, url, None).await
        })
    }

    fn post_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
        body: serde_json::Value,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>> {
        let url = Self::url(base_url, path);
        let client = self.client.clone();
        Box::pin(async move {
            Self::request(client, reqwest::Method::POST, url, Some(body)).await
        })
    }

    fn delete_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>> {
        let url = Self::url(base_url, path);
        let client = self.client.clone();
        Box::pin(async move {
            Self::request(client, reqwest::Method::DELETE, url, None).await
        })
    }

    fn put_json<'a>(
        &'a self,
        base_url: &'a str,
        path: &'a str,
        body: serde_json::Value,
    ) -> Pin<Box<dyn Future<Output = Result<serde_json::Value, SdkError>> + Send + 'a>> {
        let url = Self::url(base_url, path);
        let client = self.client.clone();
        Box::pin(async move {
            Self::request(client, reqwest::Method::PUT, url, Some(body)).await
        })
    }
}

pub fn arc_transport<T: Transport + 'static>(transport: T) -> Arc<dyn Transport> {
    Arc::new(transport)
}

File: python/medousa-sdk/src/medousa/sync/transport.py
from __future__ import annotations

import httpx

from medousa.error import SdkError
from medousa.transport import normalize_path


class SyncTransport:
    def __init__(self, bearer_token: str | None = None) -> None:
        headers = {}
        if bearer_token:
            headers["Authorization"] = f"Bearer {bearer_token}"
        self._client = httpx.Client(timeout=httpx.Timeout(120.0, connect=10.0), headers=headers)

    def close(self) -> None:
        self._client.close()

    def _url(self, base_url: str, path: str) -> str:
        return f"{base_url.rstrip('/')}{normalize_path(path)}"

    def _raise(self, response: httpx.Response) -> None:
        if response.is_success:
            return
        try:
            body = response.json()
        except Exception:
            body = response.text
        raise SdkError(
            f"HTTP request failed: {response.request.method} {response.request.url}",
            status_code=response.status_code,
            body=body,
        )

    def get_json(self, base_url: str, path: str):
        response = self._client.get(self._url(base_url, path))
        self._raise(response)
        return response.json()

    def post_json(self, base_url: str, path: str, body):
        response = self._client.post(self._url(base_url, path), json=body)
        self._raise(response)
        return response.json() if response.content else {}

    def put_json(self, base_url: str, path: str, body):
        response = self._client.put(self._url(base_url, path), json=body)
        self._raise(response)
        return response.json()

    def patch_json(self, base_url: str, path: str, body):
        response = self._client.patch(self._url(base_url, path), json=body)
        self._raise(response)
        return response.json()

    def delete_json(self, base_url: str, path: str):
        response = self._client.delete(self._url(base_url, path))
        self._raise(response)
        return response.json() if response.content else {}

    def post_empty_json(self, base_url: str, path: str):
        response = self._client.post(self._url(base_url, path))
        self._raise(response)
        return response.json() if response.content else {}
